### PR TITLE
UCP/API: error handling

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
Add error handling mode to EP parameters
Disable RNDV and ucp_tag_send_sync_nb if UCP_ERR_HANDLING_MODE_PEER is requested
Add test test_ucp_error_handling.disable_sync_send

The goal is to provide BSD-socket-like level of error reporting

TODO:
- [x] Add error handle callback API, implementation depends on #1437 
- [x] Add test on disabled RNDV

### CODE REVIEW

**Comment by shamisp** [comment]
adding @bbenton and @abouteiller

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -131,15 +131,19 @@ enum ucp_params_field {
...
+    UCP_FEATURE_RMA             = UCS_BIT(1), /**< Request remote memory
+                                                   access support */
+    UCP_FEATURE_AMO32           = UCS_BIT(2), /**< Request 32-bit atomic
+                                                   operations support */
+    UCP_FEATURE_AMO64           = UCS_BIT(3), /**< Request 64-bit atomic
+                                                   operations support */
+    UCP_FEATURE_WAKEUP          = UCS_BIT(4), /**< Request interrupt notification
+                                                   support */
+    UCP_FEATURE_FAULT_TOLERANCE = UCS_BIT(5)  /**< Request fault tolerance support,
+                                                   may disable some API returning
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: Hmmm - disabling "some" API is puzzling. What is the justification ?

Comment (shamisp) [inline src/ucp/api/ucp.h]: Also you have to elaborate what is the UCP_FEATURE_FAULT_TOLERANCE feature.

File: src/uct/api/uct.h [hunk 2]
```diff
@@ -437,6 +437,19 @@ enum {
 
 
 /**
+ * @ingroup UCT_RESOURCE
+ * @brief   @anchor UCT_FLUSH_FLAGS Flags that control completion semantic.
```
Comment (shamisp) [inline src/uct/api/uct.h]: Should include full description

File: src/uct/api/uct.h [hunk 3]
```diff
@@ -437,6 +437,19 @@ enum {
...
+ */
+enum {
+    UCT_FLUSH_FLAG_LOCAL = 0,           /**< Guarantees flush of local
+                                             outstanding operations */
```
Comment (shamisp) [inline src/uct/api/uct.h]: What is local , what is remote , etc. It has to be defined

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1392/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3321/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h (hunk 1 above)
Comment (evgeny-leksikov) [inline src/ucp/api/ucp.h]: Regarding to disabling of some APIs. Really, I see only one - ucp_tag_send_sync_nb.
The reason is no way to detect remote failure between send and acknowledge in current implementation.
This limitation is temporary because we plan to implement "keep alive" protocol in future which should resolve this limitation but we have to have it documented.

File: src/ucp/core/ucp_context.c [hunk 4]
```diff
@@ -821,6 +826,12 @@ ucs_status_t ucp_init_version(unsigned api_major_version, unsigned api_minor_ver
     ucs_queue_head_init(&context->tag.expected);
     ucs_queue_head_init(&context->tag.unexpected);
 
+    if (context->config.features & UCP_FEATURE_FAULT_TOLERANCE) {
+        ucp_tag_eager_sync_proto_p = &ucp_tag_eager_sync_unsupport_proto;
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c]: would be simple to check this in ucp_tag_send_sync_nb - since it's not fast-path call (need ack from remote) one branch should not be a problem

File: src/ucp/rma/basic_rma.c [hunk 5]
```diff
@@ -423,20 +423,25 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_worker_fence, (worker), ucp_worker_h worker)
...
+    unsigned flags = 0;
 
     UCP_THREAD_CS_ENTER_CONDITIONAL(&worker->mt_lock);
 
     while (worker->stub_pend_count > 0) {
         ucp_worker_progress(worker);
     }
 
+    if (worker->context->config.features & UCP_FEATURE_FAULT_TOLERANCE) {
+        flags |= UCT_FLUSH_FLAG_ACTIVE;
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: this is not necessarily true - flush guarantees completion of previous RMA/AMO, not the current state of remote side. We can extend the UCP flush API with flags, or add another UCP call for "keepalive", for that purpose.

File: src/ucp/tag/tag_send.c [hunk 6]
```diff
@@ -78,7 +78,7 @@ static ucs_status_t ucp_tag_req_start(ucp_request_t *req, size_t count,
                (length >= rndv_rma_thresh)) ||
                (length >= rndv_am_thresh)) && !is_iov) {
         /* RMA/AM rendezvous */
-        ucp_tag_send_start_rndv(req);
+        req->send.uct.func = proto->start_rndv;
```
Comment (yosefe) [inline src/ucp/tag/tag_send.c]: can also avoid indirect call here - disable rndv by enlarging threshold to maximum in ucp_ep_config_init, in case err handling is required

File: src/uct/api/uct.h [hunk 7]
```diff
@@ -437,6 +437,19 @@ enum {
 
 
 /**
+ * @ingroup UCT_RESOURCE
+ * @brief   @anchor UCT_FLUSH_FLAGS Flags that control completion semantic.
+ */
+enum {
+    UCT_FLUSH_FLAG_LOCAL = 0,           /**< Guarantees flush of local
```
Comment (yosefe) [inline src/uct/api/uct.h]: this change should be part of another pr - "Add active flush to UCT". That PR would be the basis of implementing keepalive mechanism in UCP.

Comment (evgeny-leksikov) [inline src/uct/api/uct.h]: if you do not have objections, let's define all APIs in one place then I split the PRs.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1410/ for details.

File: src/ucp/api/ucp.h [hunk 8]
```diff
@@ -131,15 +131,23 @@ enum ucp_params_field {
...
+                                                   \ref UCS_ERR_UNSUPPORTED
+                                                   and affect performance. Foult
```
Comment (brminich) [inline src/ucp/api/ucp.h]: Foult->Fault

File: src/ucp/core/ucp_context.c [hunk 9]
```diff
@@ -19,6 +20,10 @@
 #include <string.h>
 
 
+extern const ucp_proto_t ucp_tag_eager_sync_proto;
+extern const ucp_proto_t ucp_tag_eager_sync_unsupport_proto;
+extern const ucp_proto_t *ucp_tag_eager_sync_proto_p;
```
Comment (brminich) [inline src/ucp/core/ucp_context.c]: I'd add this pointer to the context instead. In theory we may have several contexts and not all of them may need FAULT_TOLERANCE feature

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3335/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1415/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3342/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1419/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3345/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1420/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3346/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1422/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3348/ for details (Mellanox internal link).

File: src/ucp/api/ucp_def.h [hunk 10]
```diff
@@ -87,6 +87,35 @@ typedef struct ucp_address               ucp_address_t;
...
+    UCP_ERR_HANDLING_MODE_PEER,             /**< In addition to @ref
+                                             *   UCP_ERR_HANDLING_MODE_DEFAULT
+                                             *   propagates errors also related
+                                             *   with peer failures, disables
+                                             *   protocols and APIs which may
+                                             *   cause a hang or undefined
+                                             *   behaviour in case of peer failure,
+                                             *   may affect performance and
+                                             *   memory footprint */
+    UCP_ERR_HANDLING_MODE_PEER_KEEP_ALIVE   /**< Provides similar with @ref
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: if not supported - let's not add this mode to API now

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1430/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3355/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp @bbenton pls review as well

**Comment by shamisp** [comment]
@abouteiller please review as well

File: src/ucp/api/ucp.h [hunk 11]
```diff
@@ -1680,6 +1689,8 @@ ucs_status_ptr_t ucp_tag_send_nb(ucp_ep_h ep, const void *buffer, size_t count,
  *
  * @note The user should not modify any part of the @a buffer after this
  *       operation is called, until the operation completes.
+ * @note Returns @ref UCS_ERR_UNSUPPORTED if @ref UCP_FEATURE_FAULT_TOLERANCE is
+ *       enabled.
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: Please add this is a temporary implementation related constrain that will be addressed in future releases.

Comment (evgeny-leksikov) [inline src/ucp/api/ucp.h]: Thanks!
+replace UCP_FEATURE_FAULT_TOLERANCE with UCP_ERR_HANDLING_MODE_PEER

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3362/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1435/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1436/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3363/ for details (Mellanox internal link).

File: src/ucp/api/ucp_def.h [hunk 12]
```diff
@@ -87,6 +87,29 @@ typedef struct ucp_address               ucp_address_t;
...
+ * @ingroup UCP_ENDPOINT
+ * @brief Error handling mode for the UCP endpoint.
+ * 
+ * Specifies error handling mode for the UCP endpoint.
+ */
+typedef enum {
+    UCP_ERR_HANDLING_MODE_DEFAULT,          /**< Propagates only locally
+                                             *   detected errors, provides
+                                             *   minimal overhead from
+                                             *   perfromance perspective */
```
Comment (bbenton) [inline src/ucp/api/ucp_def.h]: "provides minimal overhead" -> "imposes minimal overhead"
"from perfromance perspective" -> "from a performance perspective"

**Comment by abouteiller** [comment]
I do not think it should disable interfaces and protocols but refuse to setup the endpoint if the implementation cannot support the error mode.

I am not exactly sure what is the difference between the default and the Peer error mode: if I do a tag_send with PEER, then that send is guaranteed to report that the peer has failed (with waiting for an ack?). That is expensive and is not used in higher level abstractions. We are generally looking at a packet centric interface that guarantees generally reusability/availability of input/output buffers and not stream-like semantic with keep alive.

In the default mode, what errors are reported when a peer has failed?

File: src/ucp/api/ucp_def.h [hunk 13]
```diff
@@ -87,6 +87,29 @@ typedef struct ucp_address               ucp_address_t;
...
+    UCP_ERR_HANDLING_MODE_PEER              /**< In addition to @ref
+                                             *   UCP_ERR_HANDLING_MODE_DEFAULT
+                                             *   propagates errors also related
```
Comment (bbenton) [inline src/ucp/api/ucp_def.h]: "UCP_ERR_HANDLING_MODE_DEFAULT propagates errors also related with" ->
"UCP_ERR_HANDLING_MODE_DEFAULT, also propagates errors related to"

File: src/ucp/api/ucp_def.h [hunk 14]
```diff
@@ -87,6 +87,29 @@ typedef struct ucp_address               ucp_address_t;
...
+                                             *   with peer failures, disables
+                                             *   protocols and APIs which may
+                                             *   cause a hang or undefined
+                                             *   behaviour in case of peer failure,
```
Comment (bbenton) [inline src/ucp/api/ucp_def.h]: behaviour -> behavior (use American English spelling rather than British English spelling)

File: src/ucp/api/ucp_def.h [hunk 15]
```diff
@@ -87,6 +87,29 @@ typedef struct ucp_address               ucp_address_t;
 
 
 /**
+ * @ingroup UCP_ENDPOINT
+ * @brief Error handling mode for the UCP endpoint.
+ * 
+ * Specifies error handling mode for the UCP endpoint.
+ */
+typedef enum {
+    UCP_ERR_HANDLING_MODE_DEFAULT,          /**< Propagates only locally
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: maybe   UCP_ERR_HANDLING_MODE_NONE - no guarantees about error reporting

File: src/ucp/api/ucp_def.h [hunk 16]
```diff
@@ -87,6 +87,29 @@ typedef struct ucp_address               ucp_address_t;
...
+                                             *   perfromance perspective */
+    UCP_ERR_HANDLING_MODE_PEER              /**< In addition to @ref
+                                             *   UCP_ERR_HANDLING_MODE_DEFAULT
+                                             *   propagates errors also related
+                                             *   with peer failures, disables
+                                             *   protocols and APIs which may
+                                             *   cause a hang or undefined
+                                             *   behaviour in case of peer failure,
+                                             *   may affect performance and
+                                             *   memory footprint */
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: guarantees that send requests are always completed even in case of remote failure

**Comment by shamisp** [comment]
bot:mlx:retest

**Comment by shamisp** [comment]
bot:mlx:retest

**Comment by yosefe** [comment]
bot:bgate:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1440/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3367/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3368/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@abouteiller In general, this solution covers:
1) rndv and sync_send cases (by disabling) when sender side is waiting response from receive naturally, it may cause never completed send request if receiver was failed, lets say, during rdma-read operation for rndv case. This is temporary limitation.
2) propagating of detected peer failure from UCT level to user thru error handling callback to notificate which EP is failed

File: src/ucp/api/ucp.h [hunk 17]
```diff
@@ -165,7 +165,12 @@ enum ucp_worker_params_field {
  * present. It is used for the enablement of backward compatibility support.
  */
 enum ucp_ep_params_field {
-    UCP_EP_PARAM_FIELD_REMOTE_ADDRESS  = UCS_BIT(0)  /**< Address of remote peer */
+    UCP_EP_PARAM_FIELD_REMOTE_ADDRESS    = UCS_BIT(0), /**< Address of remote
+                                                            peer */
+    UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE = UCS_BIT(1), /**< Error handling mode.
+                                                            @ref ucp_err_handling_mode_t */
+    UCP_EP_PARAM_FIELD_ERR_HANDLE        = UCS_BIT(2)  /**< Handle to process
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: UCP_EP_PARAM_FIELD_ERR_HANDLE**R**

File: src/ucp/api/ucp.h [hunk 18]
```diff
@@ -663,6 +668,17 @@ typedef struct ucp_ep_params {
      * specified.
      */
     const ucp_address_t     *address;
+
+    /**
+     * Desired error handling mode, optional parameter. Default value is
+     * @ref UCP_ERR_HANDLING_MODE_DEFAULT
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: UCP_ERR_HANDLING_MODE_NONE

File: src/ucp/api/ucp.h [hunk 19]
```diff
@@ -663,6 +668,17 @@ typedef struct ucp_ep_params {
...
+     */
+    ucp_err_handling_mode_t err_mode;
+
+    /**
+     * Handle to process transport level failure.
+     */
+    ucp_err_handle_t        err_handle;
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: err_handler

File: src/ucp/api/ucp_def.h [hunk 20]
```diff
@@ -250,6 +272,33 @@ typedef void (*ucp_request_cleanup_callback_t)(void *request);
...
+typedef void (*ucp_err_handler_t)(void *arg, ucp_ep_h ep, ucs_status_t status);
+
+
+ /**
+ * @ingroup UCP_COMM
+ * @brief UCP endpoint error handling context.
+ * 
+ * This structure should be initialized in @ref ucp_ep_params_t to handle peer failure
+ */
+typedef struct ucp_err_handle {
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: ucp_err_handler

File: src/ucp/api/ucp_def.h [hunk 21]
```diff
@@ -250,6 +272,33 @@ typedef void (*ucp_request_cleanup_callback_t)(void *request);
...
+    ucp_err_handler_t err_handler;        /**< Error handler callback */
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: s/err_handler/cb/

File: src/ucp/api/ucp_def.h [hunk 22]
```diff
@@ -250,6 +272,33 @@ typedef void (*ucp_request_cleanup_callback_t)(void *request);
...
+    void              *err_handler_arg;   /**< User defined argument */
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: s/err_handler_arg/arg/

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1455/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3379/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1460/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3383/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
Please review the failures.
@bbenton please take another looks since docs were changed a bit

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1472/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3394/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@bbenton , @shamisp is it ok for merge?

File: src/ucp/api/ucp_def.h [hunk 23]
```diff
@@ -87,6 +87,28 @@ typedef struct ucp_address               ucp_address_t;
...
+ * 
+ * Specifies error handling mode for the UCP endpoint.
+ */
+typedef enum {
+    UCP_ERR_HANDLING_MODE_NONE,             /**< No guarantees about error
+                                             *   reporting, imposes minimal
+                                             *   overhead from a performance
+                                             *   perspective */
+    UCP_ERR_HANDLING_MODE_PEER              /**< Guarantees that send requests
+                                             *   are always completed even in
```
Comment (shamisp) [inline src/ucp/api/ucp_def.h]: Minor clarification that I would suggest - one can read completed as completed successfully. I would suggest to be more explicit ``` completed (successfully or error)```

**Comment by shamisp** [comment]
Looks good for me. @bbenton please take a look.

File: src/ucp/api/ucp.h [hunk 24]
```diff
@@ -1680,6 +1696,9 @@ ucs_status_ptr_t ucp_tag_send_nb(ucp_ep_h ep, const void *buffer, size_t count,
  *
  * @note The user should not modify any part of the @a buffer after this
  *       operation is called, until the operation completes.
+ * @note Returns @ref UCS_ERR_UNSUPPORTED if @ref UCP_ERR_HANDLING_MODE_PEER is
+ *       enabled. This is a temporary implementation related constrain that will
+ *       be addressed in future releases.
```
Comment (bbenton) [inline src/ucp/api/ucp.h]: s/implementation related constrain/implementation-related constraint/

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1493/ for details.

File: src/ucp/api/ucp_def.h [hunk 25]
```diff
@@ -250,6 +273,33 @@ typedef void (*ucp_request_cleanup_callback_t)(void *request);
...
+ /**
+ * @ingroup UCP_COMM
+ * @brief Callback to process peer failure.
+ *
+ * This callback routine is invoked when transport level error detected.
+ *
+ * @param [in]  arg      User argument to be passed to the callback.
+ * @param [in]  ep       Endpoint to handle transport level error,
+ *                       @a ep becomes unusable. All following operation on @a ep
+ *                       will return @a status error.
```
Comment (bbenton) [inline src/ucp/api/ucp_def.h]: Please update with wording similar to what we used in uct_def.h for #1437:
Endpoint to handle transport level error.  Upon return from the callback, this @a ep is no longer usable and all subsequent operations on this @a ep will fail with the error code passed in @a status

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3415/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1497/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3419/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
Failure is not related to this PR, see issue https://github.com/openucx/ucx/issues/1376

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by bbenton** [comment]
:+1: for Doc/API

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3422/ for details (Mellanox internal link).
